ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java

```diff
@@ -208,23 +228,130 @@ public boolean equals(Object obj) {
+            .map(MessageInfo::getStoreKey)
+            .collect(Collectors.toList());
+        storeKeyConverter.dropCache();
+        Map<StoreKey, StoreKey> remoteKeyToLocalKeyMap = storeKeyConverter.convert(storeKeysToConvert);
+
+        if (messagesWrittenToStore != null) {
+          // Check missing store messages in the list of messages newly written to store
+          Set<StoreKey> keysWrittenToStore =
+              messagesWrittenToStore.stream().map(MessageInfo::getStoreKey).collect(Collectors.toSet());
+          for (MessageInfo messageInfo : exchangeMetadataResponse.missingStoreMessages) {
+            if (keysWrittenToStore.contains(messageInfo.getStoreKey())) {
+              messagesFoundInStore.add(messageInfo);
+            }
+          }
+        } else {
+          // This is hit when we call compareAndRemoveMissingStoreMessages() from setExchangeMetadataResponse(). It is
+          // possible that missing store messages in exchangeMetadataResponse are written by other replica threads
+          // in parallel between the time 'exchangeMetadataResponse' is calculated and set. So, we check the local store
+          // again to find if any missing messages are written.
+          Map<MessageInfo, StoreKey> remoteMessageToConvertedKeyNonNull = new HashMap<>();
+          for (MessageInfo messageInfo : exchangeMetadataResponse.missingStoreMessages) {
+            StoreKey convertedKey = remoteKeyToLocalKeyMap.get(messageInfo.getStoreKey());
+            if (convertedKey != null) {
+              remoteMessageToConvertedKeyNonNull.put(messageInfo, convertedKey);
+            }
+          }
+          Set<StoreKey> convertedMissingStoreKeys =
+              localStore.findMissingKeys(new ArrayList<>(remoteMessageToConvertedKeyNonNull.values()));
+          remoteMessageToConvertedKeyNonNull.forEach((messageInfo, convertedKey) -> {
+            if (!convertedMissingStoreKeys.contains(convertedKey)) {
+              messagesFoundInStore.add(messageInfo);
+            }
+          });
+        }
+
+        // Go through the messages that are now found in store and reconcile delete, ttl-update and un-delete states.
+        // After that, delete them from the missingStoreMessages set and move the token forward if all missing messages
+        // are found.
+        for (MessageInfo messageInfo : messagesFoundInStore) {
+          BlobId localStoreKey = (BlobId) remoteKeyToLocalKeyMap.get(messageInfo.getStoreKey());
+
+          // 1. compare blob metadata of newly written message with the cached remote message info and
+          // reconcile delete, ttl-update and un-delete states.
+          if (localStoreKey != null) {
+            replicaThread.applyUpdatesOnLocalStoreKey(messageInfo, this, localStoreKey);
+          }
+
+          // 2. remove found message from the missing set
+          exchangeMetadataResponse.missingStoreMessages.remove(messageInfo);
```

Comment:
It seems like in a given metadata response, if there are several messages associated with same key in remote node, we get a consolidated/merged message info reflecting its latest value. For example, if we find both PUT and DELETE for same blob id, we only get DELETE. Similarly, if we find PUT and TTLUPDATE, we get one message info with TTL field updated to PUT message. Please correct me in case I am wrong here.

Even if above is not true and we receive several messages associated with same key, we might call 'applyUpdatesToBlobInLocalStore' twice which I think might still be fine as we are applying ttlupdate/delete/undelete based on latest state of local blob. 

Just wanted to note that I just separated code to apply ttlupdate/delete/undelete that we do currently in processReplicaMetadataResponse() when we find a key in local store during metadata exchange to the method 'applyUpdatesToBlobInLocalStore()' in order to reapply the same logic when we find key later in this case.